File: asset_customizations/asset_modification/customizations/asset/asset.py
from asset_customizations.asset_modification.customizations.asset.depreciation_override import scrap_asset
import frappe
from frappe.utils import getdate, today


@frappe.whitelist()
def asset_scrap_date_validation(asset_name, scrap_date, purchase_date, calculate_depreciation):
    scrap_date = getdate(scrap_date)
    today_date = getdate(today())
    purchase_date = getdate(purchase_date)
    
    if scrap_date > today_date:
        frappe.throw("Future Date Is Not Allowed")
    elif scrap_date < purchase_date:
        frappe.throw("Scrap Date Cannot Be Before Purchase Date")

    if calculate_depreciation == "0" and scrap_date >= purchase_date:
        scrap_asset(asset_name, scrap_date)
        return

    depriciation_list = frappe.db.get_all("Asset Depreciation Schedule", {"asset": asset_name}, pluck="name")
    for depriciation in depriciation_list:
        asset_depr_schedule_list = frappe.db.get_all(
            "Depreciation Schedule", 
            filters={"parent": depriciation}, 
            fields=["schedule_date", "journal_entry"], 
            order_by="schedule_date"
        )
        
        last_depreciated_row = None
        next_depriciating_row = None
        for schedule in asset_depr_schedule_list:
            if schedule["journal_entry"]:
                last_depreciated_row = schedule
            elif not schedule["journal_entry"]:
                next_depriciating_row = schedule
                break
            
        if next_depriciating_row and last_depreciated_row:
            if scrap_date < last_depreciated_row["schedule_date"] and scrap_date > purchase_date:
                frappe.throw("Asset cannot be scrap before the depreciation entry")
            elif (scrap_date <= next_depriciating_row["schedule_date"] and 
                scrap_date > last_depreciated_row["schedule_date"]):
                scrap_asset(asset_name, scrap_date)
                return
        elif not last_depreciated_row and not next_depriciating_row["journal_entry"]:
            scrap_asset(asset_name, scrap_date)
            return

File: asset_customizations/asset_modification/customizations/asset/asset.js
frappe.provide("erpnext.asset");

frappe.ui.form.on("Asset", {
    refresh: function (frm) {
		if (frm.doc.docstatus == 1) {
			if (in_list(["Submitted", "Partially Depreciated", "Fully Depreciated"], frm.doc.status)) {
				frm.add_custom_button(
					__("Scrap Asset"),
					function () {
						erpnext.asset.scrap_asset(frm);
					},
					__("Manage")
				);
            }
        }
	},
});

erpnext.asset.scrap_asset = function (frm) {
    var d = new frappe.ui.Dialog({
        title: 'Enter Date',
        fields: [
            {
                label: __("Select the date"),
                fieldname: "scrap_date",
                fieldtype: "Date",
                reqd: 1
            }
        ],
        size: 'small',
        primary_action_label: 'Submit',
        primary_action(values) {     
            frappe.call({
                args: {
                    "asset_name":frm.doc.name,
                    "scrap_date": values.scrap_date,
                    "purchase_date": frm.doc.purchase_date
                },
                method: "asset_customizations.asset_modification.customizations.asset.asset.asset_scrap_date_validation",
                callback: function(r) {
                    cur_frm.reload_doc();
                    d.hide();
                }
            })
        }
    });
    d.show();
};
